### PR TITLE
Fix merge same group expression error

### LINKED ISSUE #5205: Query returns error
### Steps to reproduce the behavior (Required)

1. CREATE EXTERNAL RESOURCE "hive_resource_emr" PROPERTIES ("type"="hive", "hive.metastore.uris"="thrift://172.26.194.238:9083")
2. CREATE TABLE `colocate_table_2` (
  `k1` datetime NOT NULL COMMENT "",
  `k2` int(11) NOT NULL COMMENT "",
  `v1` double SUM NOT NULL COMMENT ""
) ENGINE=OLAP
AGGREGATE KEY(`k1`, `k2`)
COMMENT "OLAP"
DISTRIBUTED BY HASH(`k2`) BUCKETS 8
PROPERTIES (
"replication_num" = "3",
"colocate_with" = "group1",
"in_memory" = "false",
"storage_format" = "DEFAULT"
);
3. CREATE EXTERNAL TABLE `hive_ssb_100g_orc_customer` (
  `c_custkey` int(11) NULL COMMENT "",
  `c_name` varchar(26) NULL COMMENT "",
  `c_address` varchar(41) NULL COMMENT "",
  `c_city` varchar(11) NULL COMMENT "",
  `c_nation` varchar(16) NULL COMMENT "",
  `c_region` varchar(13) NULL COMMENT "",
  `c_phone` varchar(16) NULL COMMENT "",
  `c_mktsegment` varchar(11) NULL COMMENT ""
) ENGINE=HIVE
COMMENT "PARTITION BY ()"
PROPERTIES (
"database" = "ssb_orc",
"table" = "customer",
"resource" = "hive_resource_emr",
"hive.metastore.uris"  =  "thrift://172.26.194.238:9083"
);
4. CREATE TABLE `tpch_100g_supplier` (
  `S_SUPPKEY` int(11) NOT NULL COMMENT "",
  `S_NAME` char(25) NOT NULL COMMENT "",
  `S_ADDRESS` varchar(40) NOT NULL COMMENT "",
  `S_NATIONKEY` int(11) NOT NULL COMMENT "",
  `S_PHONE` char(15) NOT NULL COMMENT "",
  `S_ACCTBAL` decimal64(15, 2) NOT NULL COMMENT "",
  `S_COMMENT` varchar(101) NOT NULL COMMENT ""
) ENGINE=OLAP
DUPLICATE KEY(`S_SUPPKEY`)
COMMENT "OLAP"
DISTRIBUTED BY HASH(`S_SUPPKEY`) BUCKETS 12
PROPERTIES (
"replication_num" = "1",
"in_memory" = "false",
"storage_format" = "DEFAULT"
);
5. CREATE EXTERNAL TABLE `hive_ssb_100g_csv_supplier` (
  `s_suppkey` int(11) NULL COMMENT "",
  `s_name` varchar(26) NULL COMMENT "",
  `s_address` varchar(26) NULL COMMENT "",
  `s_city` varchar(11) NULL COMMENT "",
  `s_nation` varchar(16) NULL COMMENT "",
  `s_region` varchar(13) NULL COMMENT "",
  `s_phone` varchar(16) NULL COMMENT ""
) ENGINE=HIVE
COMMENT "PARTITION BY ()"
PROPERTIES (
"database" = "ssb_100g_csv",
"table" = "supplier",
"resource" = "hive_resource_emr",
"hive.metastore.uris"  =  "thrift://172.26.194.238:9083"
);
6. CREATE TABLE `tpcds_100g_inventory` (
  `inv_date_sk` int(11) NOT NULL COMMENT "",
  `inv_item_sk` int(11) NOT NULL COMMENT "",
  `inv_warehouse_sk` int(11) NOT NULL COMMENT "",
  `inv_quantity_on_hand` int(11) NULL COMMENT ""
) ENGINE=OLAP
DUPLICATE KEY(`inv_date_sk`, `inv_item_sk`, `inv_warehouse_sk`)
COMMENT "OLAP"
DISTRIBUTED BY HASH(`inv_date_sk`, `inv_item_sk`, `inv_warehouse_sk`) BUCKETS 5
PROPERTIES (
"replication_num" = "1",
"in_memory" = "false",
"storage_format" = "DEFAULT"
);
7. select   ref_1.c_custkey as c0,   ref_1.c_mktsegment as c1 from   sqlsmith.colocate_table_2 as ref_0   inner join sqlsmith.hive_ssb_100g_orc_customer as ref_1 on (ref_0.k2 = ref_1.c_custkey)   left join sqlsmith.tpch_100g_supplier as ref_2 on (ref_0.k2 = ref_2.S_SUPPKEY)   right join sqlsmith.hive_ssb_100g_csv_supplier as ref_3 on (ref_1.c_name = ref_3.s_name)   right
join sqlsmith.tpcds_100g_inventory as ref_4 on (ref_0.k2 = ref_4.inv_date_sk) where   (     case       when ref_1.c_nation is
not NULL then random()       else random()     end > ref_2.S_ACCTBAL   )   and (     ref_0.v1 <= cast(       coalesce(ref_0.v1, ref_2.S_ACCTBAL) as DOUBLE     )   ) limit   63;

### Expected behavior (Required)
Query returns correctly

### Real behavior (Required)
`ERROR 1064 (HY000): Index: 0, Size: 0`
```
2022-04-18 15:34:17,214 WARN (starrocks-mysql-nio-pool-4|212) [StmtExecutor.execute():470] execute Exception, sql select   ref_1.c_custkey as c0,   ref_1.c_mktsegment as c1 from   sqlsmith.colocate_table_2 as ref_0   inner join sqlsmith.hive_ssb_100g_orc_customer as ref_1 on (ref_0.k2 = ref_1.c_custkey)   left join sqlsmith.tpch_100g_supplier as ref_2 on (ref_0.k2 = ref_2.S_SUPPKEY)   right join sqlsmith.hive_ssb_100g_csv_supplier as ref_3 on (ref_1.c_name = ref_3.s_name)   right join sqlsmith.tpcds_100g_inventory as ref_4 on (ref_0.k2 = ref_4.inv_date_sk) where   (     case       when ref_1.c_nation is not NULL then random()       else random()     end > ref_2.S_ACCTBAL   )   and (     ref_0.v1 <= cast(       coalesce(ref_0.v1, ref_2.S_ACCTBAL) as DOUBLE     )   ) limit   63
java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
        at java.util.ArrayList.rangeCheck(ArrayList.java:659) ~[?:1.8.0_292]
        at java.util.ArrayList.get(ArrayList.java:435) ~[?:1.8.0_292]
        at com.starrocks.sql.optimizer.Group.getFirstLogicalExpression(Group.java:91) ~[starrocks-fe.jar:?]
        at com.starrocks.sql.optimizer.rule.Binder.extractGroupExpression(Binder.java:131) ~[starrocks-fe.jar:?]
        at com.starrocks.sql.optimizer.rule.Binder.match(Binder.java:93) ~[starrocks-fe.jar:?]
        at com.starrocks.sql.optimizer.rule.Binder.match(Binder.java:93) ~[starrocks-fe.jar:?]
        at com.starrocks.sql.optimizer.rule.Binder.next(Binder.java:69) ~[starrocks-fe.jar:?]
        at com.starrocks.sql.optimizer.task.ApplyRuleTask.execute(ApplyRuleTask.java:63) ~[starrocks-fe.jar:?]
        at com.starrocks.sql.optimizer.task.SeriallyTaskScheduler.executeTasks(SeriallyTaskScheduler.java:42) ~[starrocks-fe.jar:?]
        at com.starrocks.sql.optimizer.Optimizer.memoOptimize(Optimizer.java:129) ~[starrocks-fe.jar:?]
        at com.starrocks.sql.optimizer.Optimizer.optimize(Optimizer.java:92) ~[starrocks-fe.jar:?]
        at com.starrocks.sql.StatementPlanner.createQueryPlan(StatementPlanner.java:84) ~[starrocks-fe.jar:?]
        at com.starrocks.sql.StatementPlanner.plan(StatementPlanner.java:54) ~[starrocks-fe.jar:?]
        at com.starrocks.qe.StmtExecutor.execute(StmtExecutor.java:309) ~[starrocks-fe.jar:?]
        at com.starrocks.qe.ConnectProcessor.handleQuery(ConnectProcessor.java:283) ~[starrocks-fe.jar:?]
        at com.starrocks.qe.ConnectProcessor.dispatch(ConnectProcessor.java:430) ~[starrocks-fe.jar:?]
        at com.starrocks.qe.ConnectProcessor.processOnce(ConnectProcessor.java:666) ~[starrocks-fe.jar:?]
        at com.starrocks.mysql.nio.ReadListener.lambda$handleEvent$457(ReadListener.java:55) ~[starrocks-fe.jar:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_292]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_292]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_292]
```

### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
branch-2.2 release 79e412e

### PR DESCRIPTION
**Pull request by Seaven**
## What type of PR is this：
- [x] bug
- [ ] feature
- [ ] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #5205 #5277

## Problem Summary(Required) ：

In memo merge phase, when we move the group expression of delete group to target group, and there has the same group expression in the target group, we will remove error group expression finally

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 26 / 26 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/Memo.java|26|26|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 28 / 29 (96.55%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/Group.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/Memo.java|28|28|100.00%|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Memo.java [hunk 1]
```diff
@@ -159,51 +159,62 @@ private void mergeGroup(Group srcGroup, Group dstGroup) {
...
         }
 
-        for (GroupExpression groupExpression : needReinsertedExpressions) {
-            if (!groupExpressions.containsKey(groupExpression)) {
-                groupExpressions.put(groupExpression, groupExpression);
+        for (GroupExpression reinsertExpression : needReinsertedExpressions) {
+            // reinsert maybe in groupExpressions because his input was modify
+            if (!groupExpressions.containsKey(reinsertExpression)) {
+                groupExpressions.put(reinsertExpression, reinsertExpression);
+                reinsertExpression.getGroup().addExpression(reinsertExpression);
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Memo.java:204]: would better add Preconditions.checkState in group addExpression

**Review by kangkaisen — APPROVED** [review]

**Comment by kangkaisen** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 30 / 31 (96.77%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/Group.java|2|3|66.67%|
|:large_blue_circle:|com/starrocks/sql/optimizer/Memo.java|28|28|100.00%|

**Review by Youngwb — APPROVED** [review]

**Comment by Seaven** [comment]
@mergeify backport branch-2.2

**Comment by Seaven** [comment]
@Mergeify backport branch-2.2

**Comment by Seaven** [comment]
@Mergeifyio backport branch-2.2

**Comment by Seaven** [comment]
@mergeifyio backport branch-2.2

**Comment by Seaven** [comment]
@Mergifyio backport branch-2.2

**Pull request merged**